lib/jackalope/work_list.ex

```diff
@@ -1,128 +1,73 @@
 defmodule Jackalope.WorkList do
-  @moduledoc false
-
-  require Logger
-  @default_max_size 100
-
-  defstruct items: [],
-            pending: %{},
-            max_size: @default_max_size
-
-  @type t() :: %__MODULE__{items: list(), max_size: non_neg_integer()}
-
-  @spec new(list(), non_neg_integer()) :: t()
-  def new(items, max_size \\ @default_max_size) when max_size > 0 do
-    %__MODULE__{items: items, max_size: max_size}
-  end
-
-  @spec push(t(), any) :: t()
-  def push(work_list, item) do
-    updated_items =
-      [item | work_list.items]
-      |> bound(work_list.max_size)
-
-    %__MODULE__{work_list | items: updated_items}
-  end
-
-  @spec prepend(t(), list) :: t()
-  def prepend(work_list, items) when is_list(items) do
-    updated_items =
-      (items ++ work_list.items)
-      |> bound(work_list.max_size)
-
-    %__MODULE__{work_list | items: updated_items}
+  @moduledoc """
+  Behaviour for work lists and dispatcher to implementations.
+  """
+
+  @type t() :: any()
+  @type item() :: any()
+  @type work_list :: {module, t}
+  @type expiration :: {integer, non_neg_integer | :infinity}
+
+  @callback new(function(), function(), non_neg_integer()) :: t()
+  @callback push(t(), any) :: t()
+  @callback peek(t()) :: item()
+  @callback pop(t()) :: t()
+  @callback pending(t(), reference()) :: t()
+  @callback reset_pending(t()) :: t()
+  @callback done(t(), reference()) :: {t(), item()}
+  @callback count(t()) :: non_neg_integer()
+  @callback count_pending(t()) :: non_neg_integer()
+  @callback empty?(t()) :: boolean
+  @callback remove_all(t()) :: t()
+
+  @spec now :: integer
+  def now(), do: System.monotonic_time(:millisecond)
+
+  @spec unexpired?(any, (any -> expiration)) :: boolean
+  def unexpired?(item, expiration_fn), do: not after?(expiration(0), expiration_fn.(item))
+
+  @spec expiration(integer) :: expiration
+  def expiration(ttl), do: {now(), ttl}
+
+  @doc "Does the first expiration come after the second?"
+  @spec after?(expiration, expiration) :: boolean
+  def after?({_t1, :infinity}, _), do: true
+  def after?(_, {_t2, :infinity}), do: false
+  def after?({t1, delta1}, {t2, delta2}), do: t1 + delta1 > t2 + delta2
+
+  @doc "Recalculate an old expiration given the latest time before stop and the earliest time after restart.
+  Assumes restart happened soon after stopping."
+  @spec rebase_expiration(expiration, integer, integer) :: expiration
+  def rebase_expiration({_t, :infinity}, _stop_time, restart_time), do: {restart_time, :infinity}
+
+  def rebase_expiration({t, ttl}, stop_time, restart_time) do
+    time_passed = stop_time - t
+    {restart_time, max(0, ttl - time_passed)}
```

Comment:
I think that keeping time as a tuple isn't needed. This code looks the same as if you'd store the timeout time for each item. The rebase operation would still be to compute the delta time that passed and subtract it off. Just storing the time out times seems simpler, though.